#!/bin/bash


## help message
usage() {
echo -e "
This script uses gene sequences in one fasta file to generate tree for iTOL viewer;
    Tools used: mafft, trimAl, and IQ-TREE.  

outputs: tree file generated by iqtree

Required arguments:
-i      input gene file (in DNA format here)

Optional arguments:
-h, --help   show this help message
"
}

if [ -z "$1" ] || [[ $1 == -h ]] || [[ $1 == --help ]]; then
  usage
  exit 1
fi

## determine if the 1st argument is start with "-"
head=${1:0:1}
if [[ $head != - ]]; then
  echo "syntax error. You need an argument"
  usage
  exit 1
fi

## arguments, variables, and default values
gene_number_max=100
while getopts "i:n:s:" opt
do
  case $opt in
    i)
        input_gene=$OPTARG;;

    \?)
        echo "invalid argument"
        exit 1;;
  esac
done



## working directory
path_input=$(readlink -f $input_gene)
workDir=${path_input%/*}
cd $workDir

## file names
input_gene_name=${input_gene##*/}
input_gene_name_1=${input_gene_name%.*}


## tools
module use /fs/project/PAS1117/zhiping/software/1_modules/modulefiles
module load mafft/7.429
export PATH=/users/PAS1117/osu7810/software/trimal.v1.2rev59/trimAl/source/:$PATH
module load IQ-TREE/1.6.11

## alignment by mafft
mafft --reorder --bl 30 --op 1.0 --maxiterate 1000 --retree 1 --genafpair --quiet $input_gene > ${workDir}/${input_gene_name_1}_aligned.fa

## trim by trimA - adjust gap parameters if necessary
trimal -in ${workDir}/${input_gene_name_1}_aligned.fa -out ${workDir}/${input_gene_name_1}_aligned_trim.fa -gappyout


## iqtree - find the best-fit model and then automatically use it to plot tree
iqtree -s ${workDir}/${input_gene_name_1}_aligned_trim.fa -m MFP -bb 1000 -alrt 1000 -redo > log.txt

## end
echo "\n*** Tree path: ${workDir}/04_tree_input_seq/"
